src/core/packages/saved-objects/api-server-internal/src/lib/apis/search.ts

```diff
@@ -0,0 +1,218 @@
+      index: esOptions.pit ? undefined : commonHelper.getIndicesForTypes(types),
+      query,
+    },
+    {
+      ignore: [404],
+      meta: true,
+    }
+  );
+  if (result.statusCode === 404) {
+    if (!isSupportedEsServer(result.headers)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError();
+    }
+    // 404 is only possible here if the index is missing, which
+    // we don't want to leak, see "404s from missing index" above
+    return createEmptySearchResponse();
+  }
+
+  const redactTypeMapParams: GetFindRedactTypeMapParams = {
+    previouslyCheckedNamespaces: spacesToAuthorize,
+    objects: [],
+  };
+  for (const hit of result.body.hits.hits) {
+    if (serializer.isRawSavedObject(hit as SavedObjectsRawDoc)) {
+      redactTypeMapParams.objects.push({
+        type: hit._source!.type,
+        id: hit._id!,
+        existingNamespaces: hit._source!.namespaces ?? [],
+      });
+    }
+  }
+
+  const redactTypeMap = redactTypeMapParams
+    ? await securityExtension?.getFindRedactTypeMap(redactTypeMapParams)
+    : undefined;
+
+  // Migrations and encryption don't work on raw documents. To process them we have
+  // to serialize raw documents to saved objects and then deserialize them back to
+  // raw documents again.
+  const processHit = async (hit: estypes.SearchHit<T>): Promise<estypes.SearchHit<T>> => {
+    if (!serializer.isRawSavedObject(hit as SavedObjectsRawDoc)) {
+      return hit;
+    }
+
+    const savedObject = serializerHelper.rawToSavedObject(hit as SavedObjectsRawDoc);
+    if (options.fields?.length) {
+      // If the fields argument is set, don't migrate.
+      // This document may only contains a subset of it's fields meaning the migration
+      // (transform and forwardCompatibilitySchema) is not guaranteed to succeed. We
+      // still try to decrypt/redact the fields that are present in the document.
+      const decrypted = await encryptionHelper.optionallyDecryptAndRedactSingleResult(
+        savedObject,
+        redactTypeMap
+      );
+
+      return {
+        ...hit,
+        ...serializer.savedObjectToRaw(decrypted),
+      } as estypes.SearchHit<T>;
+    }
+
+    const migratedSavedObject = await migrationHelper.migrateAndDecryptStorageDocument({
+      document: savedObject,
+      typeMap: redactTypeMap,
+    });
+
+    return {
+      ...hit,
+      ...serializer.savedObjectToRaw(migratedSavedObject),
+    } as estypes.SearchHit<T>;
+  };
+
+  const processedHits: estypes.SearchHit<T>[] = [];
+  for (const hit of result.body.hits.hits) {
+    processedHits.push(await processHit(hit));
+  }
+  result.body.hits.hits = processedHits;
+
+  return result.body;
+}
+
+/**
+ * This is how we inject our namespaces scope into the user query. We create a
+ * nested `bool.must` clause for the user query and keep our query in a neighboring
+ * `bool.should` clause.
+ */
+export function mergeUserQueryWithNamespacesBoolForSearch(
```

Comment:
@rudolf , this is were we merge namespace and user queries... I think this is a nice simplification as we can now avoid the `deepMerge` we were doing before which makes this a bit more transparent (user input will not collide with our injected filters).